File: syncflow-java/syncflow-statistics/src/main/java/com/syncflow/statistics/controller/query/QueryController.java
package com.syncflow.statistics.controller.query;

import com.syncflow.common.result.Result;
import com.syncflow.statistics.dto.query.*;
import com.syncflow.statistics.service.query.QueryService;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * Query / statistics controller for cross-cutting queries.
 */
@RestController
@RequestMapping("/api/query")
public class QueryController {

    private final QueryService queryService;

    public QueryController(QueryService queryService) {
        this.queryService = queryService;
    }

    /**
     * Get task statistics.
     */
    @GetMapping("/task-stats")
    public Result<TaskStatsVO> getTaskStats(
            @RequestParam(required = false) Long projectId) {
        TaskStatsVO vo = queryService.getTaskStats(projectId);
        return Result.success(vo);
    }

    /**
     * Get project statistics.
     */
    @GetMapping("/project-stats")
    public Result<ProjectStatsVO> getProjectStats() {
        ProjectStatsVO vo = queryService.getProjectStats();
        return Result.success(vo);
    }

    /**
     * Get overdue tasks.
     */
    @GetMapping("/overdue-tasks")
    public Result<List<OverdueTaskVO>> getOverdueTasks(
            @RequestParam(required = false) Long projectId) {
        List<OverdueTaskVO> result = queryService.getOverdueTasks(projectId);
        return Result.success(result);
    }

    /**
     * Get project progress.
     */
    @GetMapping("/project-progress/{projectId}")
    public Result<ProjectProgressVO> getProjectProgress(@PathVariable Long projectId) {
        ProjectProgressVO vo = queryService.getProjectProgress(projectId);
        return Result.success(vo);
    }

    /**
     * Get user workload.
     */
    @GetMapping("/user-workload/{userId}")
    public Result<UserWorkloadVO> getUserWorkload(@PathVariable Long userId) {
        UserWorkloadVO vo = queryService.getUserWorkload(userId);
        return Result.success(vo);
    }

    /**
     * Get department statistics.
     */
    @GetMapping("/department-stats/{departmentId}")
    public Result<DepartmentStatsVO> getDepartmentStats(@PathVariable Long departmentId) {
        DepartmentStatsVO vo = queryService.getDepartmentStats(departmentId);
        return Result.success(vo);
    }

    /**
     * Export tasks as CSV.
     */
    @GetMapping("/export/tasks")
    public ResponseEntity<byte[]> exportTasks(
            @RequestParam(required = false) Long projectId) {
        String csv = queryService.exportTasks(projectId);
        byte[] bytes = csv.getBytes();

        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=tasks.csv")
                .contentType(MediaType.parseMediaType("text/csv"))
                .contentLength(bytes.length)
                .body(bytes);
    }

    /**
     * Export projects as CSV.
     */
    @GetMapping("/export/projects")
    public ResponseEntity<byte[]> exportProjects() {
        String csv = queryService.exportProjects();
        byte[] bytes = csv.getBytes();

        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=projects.csv")
                .contentType(MediaType.parseMediaType("text/csv"))
                .contentLength(bytes.length)
                .body(bytes);
    }
}

File: v1-nestjs/server/src/query/query.controller.ts
import { Controller, Get, Header, Query, Param } from '@nestjs/common';
import { QueryService } from './query.service';

@Controller('query')
export class QueryController {
  constructor(private readonly queryService: QueryService) {}

  @Get('task-stats')
  async getTaskStats() {
    const data = await this.queryService.getTaskStats();
    return { code: 0, data };
  }

  @Get('project-stats')
  async getProjectStats() {
    const data = await this.queryService.getProjectStats();
    return { code: 0, data };
  }

  @Get('overdue-tasks')
  async getOverdueTasks() {
    const data = await this.queryService.getOverdueTasks();
    return { code: 0, data };
  }

  @Get('project-progress/:projectId')
  async getProjectProgress(@Param('projectId') projectId: string) {
    const data = await this.queryService.getProjectProgress(projectId);
    return { code: 0, data };
  }

  @Get('user-workload/:userId')
  async getUserWorkload(@Param('userId') userId: string) {
    const data = await this.queryService.getUserWorkload(userId);
    return { code: 0, data };
  }

  @Get('department-stats/:departmentId')
  async getDepartmentStats(@Param('departmentId') departmentId: string) {
    const data = await this.queryService.getDepartmentStats(departmentId);
    return { code: 0, data };
  }

  @Get('export/tasks')
  @Header('Content-Type', 'text/csv')
  async exportTasks(
    @Query('projectId') projectId?: string,
    @Query('status') status?: string,
    @Query('priority') priority?: string,
  ) {
    return this.queryService.exportTasks({ projectId, status, priority });
  }

  @Get('export/projects')
  @Header('Content-Type', 'text/csv')
  async exportProjects(@Query('status') status?: string) {
    return this.queryService.exportProjects({ status });
  }
}
